File: solutions/golang/74/solution.go
func searchMatrix(matrix [][]int, target int) bool {
    M := len(matrix)
    N := len(matrix[0])
    left := 0
    right := N * M - 1
    for (left <= right) {
        mid := left + (right - left) / 2
        row := mid / N
        col := mid % N
        num := matrix[row][col]
        if num == target {
            return true
        } else if num > target {
            right = mid - 1
        } else {
            left = mid + 1
        }
    }
    return false
}

File: solutions/cpp/74/solution.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        const int M = matrix.size();
        const int N = matrix[0].size();
        int left = 0;
        int right = M * N - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            int row = mid / N;
            int col = mid % N;
            int num = matrix[row][col];
            if (num == target) {
                return true;
            }
            else if (num > target) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return false;
    }
};
